optuna/study/_multi_objective.py

```diff
@@ -69,6 +74,94 @@ def _get_pareto_front_trials(study: "optuna.study.Study") -> List[FrozenTrial]:
     return _get_pareto_front_trials_by_trials(study.trials, study.directions)
 
 
+def _fast_non_dominated_sort(
+    objective_values: np.ndarray,
+    *,
+    penalty: np.ndarray | None = None,
+    n_below: int | None = None,
+) -> np.ndarray:
+    if penalty is None:
+        ranks, _ = _calculate_nondomination_rank(objective_values, n_below=n_below)
+        return ranks
+
+    if len(penalty) != len(objective_values):
+        raise ValueError(
+            "The length of penalty and objective_values must be same, but got "
+            "len(penalty)={} and len(objective_values)={}.".format(
+                len(penalty), len(objective_values)
+            )
+        )
+    nondomination_rank = np.full(len(objective_values), -1)
+    is_nan = np.isnan(penalty)
+    n_below = n_below or len(objective_values)
+
+    # First, we calculate the domination rank for feasible trials.
+    is_feasible = np.logical_and(~is_nan, penalty <= 0)
+    ranks, bottom_rank = _calculate_nondomination_rank(
+        objective_values[is_feasible], n_below=n_below
+    )
+    nondomination_rank[is_feasible] += 1 + ranks
+    n_below -= np.count_nonzero(is_feasible)
+
+    # Second, we calculate the domination rank for infeasible trials.
+    is_infeasible = np.logical_and(~is_nan, penalty > 0)
+    ranks, bottom_rank = _calculate_nondomination_rank(
+        penalty[is_infeasible, np.newaxis], n_below=n_below, base_rank=bottom_rank + 1
+    )
+    nondomination_rank[is_infeasible] += 1 + ranks
+    n_below -= np.count_nonzero(is_infeasible)
+
+    # Third, we calculate the domination rank for trials with no penalty information.
+    ranks, _ = _calculate_nondomination_rank(
+        objective_values[is_nan], n_below=n_below, base_rank=bottom_rank + 1
+    )
+    nondomination_rank[is_nan] += 1 + ranks
+
+    return nondomination_rank
+
+
+def _calculate_nondomination_rank(
+    objective_values: np.ndarray,
+    *,
+    n_below: int | None = None,
+    base_rank: int = 0,
+) -> tuple[np.ndarray, int]:
+    # Normalize n_below.
+    n_below = n_below or len(objective_values)
+    n_below = min(n_below, len(objective_values))
+
+    # The ndarray `domination_matrix` is a boolean 2d matrix where
+    # `domination_matrix[i, j] == True` means that the j-th trial dominates the i-th trial in the
+    # given multi objective minimization problem.
```

Comment:
The name of variable is inconsistent.
```suggestion
    # The ndarray `domination_mat` is a boolean 2d matrix where
    # `domination_mat[i, j] == True` means that the j-th trial dominates the i-th trial in the
    # given multi objective minimization problem.
```